File: dart/packages/fory/test/container_memory_budget_test.dart
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import 'dart:typed_data';

import 'package:fory/fory.dart';
import 'package:fory/src/context/meta_string_reader.dart';
import 'package:fory/src/context/ref_reader.dart';
import 'package:fory/src/resolver/type_resolver.dart';
import 'package:fory/src/serializer/collection_serializers.dart';
import 'package:fory/src/serializer/map_serializers.dart';
import 'package:test/test.dart';

part 'container_memory_budget_test.fory.dart';

const Matcher _throwsContainerBudget = ThrowsContainerBudget();

@ForyStruct()
class BudgetGeneratedEnvelope {
  BudgetGeneratedEnvelope();

  @ListField(element: Int32Type(encoding: Encoding.fixed))
  List<int> ids = <int>[];

  @SetField(element: StringType())
  Set<String> tags = <String>{};

  @MapField(
    key: StringType(),
    value: Int32Type(encoding: Encoding.fixed),
  )
  Map<String, int> counts = <String, int>{};
}

@ForyStruct()
class BudgetCompatibleListEnvelope {
  BudgetCompatibleListEnvelope();

  @ListField(element: Int32Type(encoding: Encoding.fixed))
  List<int> values = <int>[];
}

@ForyStruct()
class BudgetCompatibleArrayEnvelope {
  BudgetCompatibleArrayEnvelope();

  @ArrayField(element: Int32Type())
  Int32List values = Int32List(0);
}

final class ThrowsContainerBudget extends Matcher {
  const ThrowsContainerBudget();

  @override
  Description describe(Description description) {
    return description.add('throws a maxContainerMemoryBytes StateError');
  }

  @override
  bool matches(Object? item, Map<Object?, Object?> matchState) {
    if (item is! Function) {
      return false;
    }
    try {
      item();
    } on StateError catch (error) {
      return error.message.contains('maxContainerMemoryBytes');
    }
    return false;
  }
}

void _registerGenerated(Fory fory) {
  ContainerMemoryBudgetTestForyModule.register(
    fory,
    BudgetGeneratedEnvelope,
    name: 'test.BudgetGeneratedEnvelope',
  );
}

void _registerCompatibleList(Fory fory) {
  ContainerMemoryBudgetTestForyModule.register(
    fory,
    BudgetCompatibleListEnvelope,
    name: 'test.BudgetCompatibleEnvelope',
  );
}

void _registerCompatibleArray(Fory fory) {
  ContainerMemoryBudgetTestForyModule.register(
    fory,
    BudgetCompatibleArrayEnvelope,
    name: 'test.BudgetCompatibleEnvelope',
  );
}

ReadContext _readContext(Buffer buffer, {int maxContainerMemoryBytes = -1}) {
  final config = Config(maxContainerMemoryBytes: maxContainerMemoryBytes);
  final resolver = TypeResolver(config);
  return ReadContext(config, resolver, RefReader(), MetaStringReader(resolver))
    ..prepare(buffer);
}

Uint8List _serialize(Object? value) => Fory().serialize(value);

Object? _readWithBudget(Object? value, int budget) {
  return Fory(
    maxContainerMemoryBytes: budget,
  ).deserialize<Object?>(_serialize(value));
}

void main() {
  group('container memory budget', () {
    test('known length auto derives from input bytes', () {
      final buffer = Buffer.wrap(Uint8List(17));
      final context = _readContext(buffer);

      expect(context.effectiveContainerMemoryBytes, equals(17 * 8 + 64 * 1024));
      expect(
        () => context.reserveContainerMemory(17 * 8 + 64 * 1024),
        returnsNormally,
      );
      expect(() => context.reserveContainerMemory(1), _throwsContainerBudget);
    });

    test('explicit config overrides auto', () {
      final buffer = Buffer.wrap(Uint8List(4096));
      final context = _readContext(buffer, maxContainerMemoryBytes: 31);

      expect(context.effectiveContainerMemoryBytes, equals(31));
      expect(() => context.reserveContainerMemory(31), returnsNormally);
      expect(() => context.reserveContainerMemory(1), _throwsContainerBudget);
      expect(() => Fory(maxContainerMemoryBytes: 0), throwsArgumentError);
      expect(() => Fory(maxContainerMemoryBytes: -2), throwsArgumentError);
    });

    test('charges nested empty containers', () {
      final value = <Object?>[<Object?>[]];

      expect(() => _readWithBudget(value, 51), _throwsContainerBudget);
      expect(_readWithBudget(value, 52), equals(value));
    });

    test('charges sibling containers cumulatively', () {
      final value = <Object?>[<Object?>[], <Object?>[], <Object?>[]];

      expect(() => _readWithBudget(value, 107), _throwsContainerBudget);
      expect(_readWithBudget(value, 108), equals(value));
    });

    test('charges map table and entries', () {
      final value = <Object?, Object?>{'a': 1};

      expect(() => _readWithBudget(value, 99), _throwsContainerBudget);
      expect(_readWithBudget(value, 100), equals(value));
    });

    test('charges generated list set and map reads', () {
      final writer = Fory();
      _registerGenerated(writer);
      final bytes = writer.serialize(
        BudgetGeneratedEnvelope()
          ..ids = <int>[1]
          ..tags = <String>{'x'}
          ..counts = <String, int>{'one': 1},
      );

      final failingReader = Fory(maxContainerMemoryBytes: 183);
      _registerGenerated(failingReader);
      expect(
        () => failingReader.deserialize<BudgetGeneratedEnvelope>(bytes),
        _throwsContainerBudget,
      );

      final passingReader = Fory(maxContainerMemoryBytes: 184);
      _registerGenerated(passingReader);
      final roundTrip = passingReader.deserialize<BudgetGeneratedEnvelope>(
        bytes,
      );
      expect(roundTrip.ids, equals(<int>[1]));
      expect(roundTrip.tags, equals(<String>{'x'}));
      expect(roundTrip.counts, equals(<String, int>{'one': 1}));
    });

    test('charges compatible list array materialization', () {
      final listWriter = Fory();
      _registerCompatibleList(listWriter);
      final listBytes = listWriter.serialize(
        BudgetCompatibleListEnvelope()..values = <int>[1, 2, 3],
      );

      final arrayFail = Fory(maxContainerMemoryBytes: 27);
      _registerCompatibleArray(arrayFail);
      expect(
        () => arrayFail.deserialize<BudgetCompatibleArrayEnvelope>(listBytes),
        _throwsContainerBudget,
      );

      final arrayPass = Fory(maxContainerMemoryBytes: 28);
      _registerCompatibleArray(arrayPass);
      expect(
        arrayPass
            .deserialize<BudgetCompatibleArrayEnvelope>(listBytes)
            .values
            .toList(),
        equals(<int>[1, 2, 3]),
      );

      final arrayWriter = Fory();
      _registerCompatibleArray(arrayWriter);
      final arrayBytes = arrayWriter.serialize(
        BudgetCompatibleArrayEnvelope()
          ..values = Int32List.fromList(<int>[1, 2, 3]),
      );

      final listFail = Fory(maxContainerMemoryBytes: 35);
      _registerCompatibleList(listFail);
      expect(
        () => listFail.deserialize<BudgetCompatibleListEnvelope>(arrayBytes),
        _throwsContainerBudget,
      );

      final listPass = Fory(maxContainerMemoryBytes: 36);
      _registerCompatibleList(listPass);
      expect(
        listPass.deserialize<BudgetCompatibleListEnvelope>(arrayBytes).values,
        equals(<int>[1, 2, 3]),
      );
    });

    test('skips strings binary and dense typed arrays', () {
      final fory = Fory(maxContainerMemoryBytes: 1);
      final text = List<String>.filled(128, 'x').join();

      expect(fory.deserialize<String>(Fory().serialize(text)), hasLength(128));
      expect(
        fory.deserialize<Uint8List>(Fory().serialize(Uint8List(128))).length,
        equals(128),
      );
      expect(
        fory.deserialize<Int32List>(Fory().serialize(Int32List(32))).length,
        equals(32),
      );
    });

    test('keeps byte availability checks before allocation', () {
      final listBuffer = Buffer()
        ..writeVarUint32(64)
        ..writeUint8(0);
      final listContext = _readContext(listBuffer);
      expect(
        () => ListSerializer.readPayload(listContext, null),
        throwsStateError,
      );

      final mapBuffer = Buffer()..writeVarUint32(64);
      final mapContext = _readContext(mapBuffer);
      expect(
        () => MapSerializer.readPayload(mapContext, null, null),
        throwsStateError,
      );
    });
  });
}

File: javascript/test/containerMemoryBudget.test.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import Fory, { Type } from '../packages/core/index';
import { describe, expect, test } from '@jest/globals';

const KNOWN_SLACK_BYTES = 64 * 1024;

function serializeAny(value: unknown) {
  return new Fory({ compatible: false, ref: true }).serialize(value);
}

function deserializeAny(bytes: Uint8Array, maxContainerMemoryBytes: number) {
  return new Fory({
    compatible: false,
    ref: true,
    maxContainerMemoryBytes,
  }).deserialize(bytes);
}

describe('container memory budget', () => {
  test('uses known length auto budget', () => {
    const inputBytes = 17;
    const fory = new Fory({ compatible: false });
    const budget = inputBytes * 8 + KNOWN_SLACK_BYTES;

    fory.readContext.reset(new Uint8Array(inputBytes));
    expect(() => fory.readContext.reserveContainerMemory(budget)).not.toThrow();
    expect(() => fory.readContext.reserveContainerMemory(1)).toThrow(
      /maxContainerMemoryBytes/,
    );
  });

  test('validates explicit config', () => {
    expect(() => new Fory({ maxContainerMemoryBytes: 0 })).toThrow(
      /maxContainerMemoryBytes/,
    );
    expect(() => new Fory({ maxContainerMemoryBytes: -2 })).toThrow(
      /maxContainerMemoryBytes/,
    );

    const fory = new Fory({ maxContainerMemoryBytes: 24 });
    fory.readContext.reset(new Uint8Array(1));
    expect(() => fory.readContext.reserveContainerMemory(0)).not.toThrow();
    expect(() => fory.readContext.reserveContainerMemory(24)).not.toThrow();
    expect(() => fory.readContext.reserveContainerMemory(1)).toThrow(
      /maxContainerMemoryBytes/,
    );
  });

  test('uses parent storage for nested empty containers', () => {
    const typeInfo = Type.struct('budget.nested.empty', {
      values: Type.list(Type.list(Type.int32({ encoding: 'fixed' }))).setId(1),
    });
    const writer = new Fory({ compatible: false, ref: true });
    const bytes = writer.register(typeInfo).serialize({ values: [[]] });
    const passingReader = new Fory({
      compatible: false,
      ref: true,
      maxContainerMemoryBytes: 4,
    }).register(typeInfo);
    const failingReader = new Fory({
      compatible: false,
      ref: true,
      maxContainerMemoryBytes: 3,
    }).register(typeInfo);

    expect(() => failingReader.deserialize(bytes)).toThrow(
      /maxContainerMemoryBytes/,
    );
    expect(passingReader.deserialize(bytes)).toEqual({ values: [[]] });
  });

  test('reserves sibling containers cumulatively', () => {
    const typeInfo = Type.struct('budget.sibling.empty', {
      values: Type.list(Type.list(Type.int32({ encoding: 'fixed' }))).setId(1),
    });
    const writer = new Fory({ compatible: false, ref: true });
    const bytes = writer.register(typeInfo).serialize({
      values: [[], [], []],
    });
    const passingReader = new Fory({
      compatible: false,
      ref: true,
      maxContainerMemoryBytes: 12,
    }).register(typeInfo);
    const failingReader = new Fory({
      compatible: false,
      ref: true,
      maxContainerMemoryBytes: 11,
    }).register(typeInfo);

    expect(() => failingReader.deserialize(bytes)).toThrow(
      /maxContainerMemoryBytes/,
    );
    expect(passingReader.deserialize(bytes)).toEqual({
      values: [[], [], []],
    });
  });

  test('reserves map entries', () => {
    const bytes = serializeAny(new Map([[1, 2]]));

    expect(() => deserializeAny(bytes, 7)).toThrow(/maxContainerMemoryBytes/);
    expect(deserializeAny(bytes, 8)).toEqual(new Map([[1, 2]]));
  });

  test('reserves generated containers', () => {
    const typeInfo = Type.struct('budget.generated', {
      list: Type.list(Type.int32({ encoding: 'fixed' })).setId(1),
      set: Type.set(Type.string()).setId(2),
      map: Type.map(Type.string(), Type.int32({ encoding: 'fixed' })).setId(3),
    });
    const writer = new Fory({ compatible: false, ref: true });
    const bytes = writer.register(typeInfo).serialize({
      list: [1],
      set: new Set(['a']),
      map: new Map([['k', 1]]),
    });
    const passingReader = new Fory({
      compatible: false,
      ref: true,
      maxContainerMemoryBytes: 16,
    }).register(typeInfo);
    const failingReader = new Fory({
      compatible: false,
      ref: true,
      maxContainerMemoryBytes: 15,
    }).register(typeInfo);

    expect(() => failingReader.deserialize(bytes)).toThrow(
      /maxContainerMemoryBytes/,
    );
    expect(passingReader.deserialize(bytes)).toEqual({
      list: [1],
      set: new Set(['a']),
      map: new Map([['k', 1]]),
    });
  });

  test('reserves compatible typed arrays', () => {
    const writerType = Type.struct(9010, {
      values: Type.list(Type.int32({ encoding: 'fixed' })).setId(1),
    });
    const readerType = Type.struct(9010, {
      values: Type.int32Array().setId(1),
    });
    const writer = new Fory({ compatible: true });
    const bytes = writer.register(writerType).serialize({ values: [1, 2, 3] });
    const passingReader = new Fory({
      compatible: true,
      maxContainerMemoryBytes: 12,
    }).register(readerType);
    const failingReader = new Fory({
      compatible: true,
      maxContainerMemoryBytes: 11,
    }).register(readerType);

    expect(() => failingReader.deserialize(bytes)).toThrow(
      /maxContainerMemoryBytes/,
    );
    expect(Array.from(passingReader.deserialize(bytes).values)).toEqual([
      1,
      2,
      3,
    ]);
  });

  test('skips scalar dense owners', () => {
    const typeInfo = Type.struct('budget.skipped', {
      text: Type.string().setId(1),
      binary: Type.binary().setId(2),
      values: Type.int32Array().setId(3),
    });
    const writer = new Fory({ compatible: false, ref: true });
    const bytes = writer.register(typeInfo).serialize({
      text: 'hello',
      binary: new Uint8Array([1, 2, 3]),
      values: new Int32Array([1, 2, 3]),
    });
    const reader = new Fory({
      compatible: false,
      ref: true,
      maxContainerMemoryBytes: 1,
    }).register(typeInfo);

    expect(reader.deserialize(bytes)).toEqual({
      text: 'hello',
      binary: new Uint8Array([1, 2, 3]),
      values: new Int32Array([1, 2, 3]),
    });
  });

  test('keeps byte checks', () => {
    const typeInfo = Type.struct('budget.bytecheck', {
      values: Type.int32Array().setId(1),
    });
    const writer = new Fory({ compatible: false, ref: true });
    const bytes = writer.register(typeInfo).serialize({
      values: new Int32Array([1, 2, 3]),
    });
    const reader = new Fory({
      compatible: false,
      ref: true,
      maxContainerMemoryBytes: 1024 * 1024,
    }).register(typeInfo);

    expect(() => reader.deserialize(bytes.slice(0, bytes.length - 1))).toThrow();
  });
});
